Plane: add Autoland mode

Hwurzburg opened the pull request:
Captures ground course when gnd speed reaches 5m/s as takeoff direction in NAV_TAKEOFF and Mode Takeoff (if a reliable way to obtain a takeoff direction in other modes is determined, this can be expanded in the future)

Currently valid only for fixed wing configuration, quadplane wont enter mode...

Upon entry creates a final approach waypoint at parameterized distance and alt from home in the opposite direction of autotakeoff direction, flies to it and does a land exactly like NAV_LAND using full landing code for aborts. (Emulates my UniversalAutoLand LUA script #28591)

~~If a DO_LAND_START sequence(s) exist, will jump to them instead~~

If takeoff direction has not been captured then it will just not enter the mode

If not flying it will not enter the mode

Adds it as a long failsafe action (will RTL if no takeoff direction has been captured)...qplane ignores it

Includes an autotest

edit: can now be used in QPlane if fw landing option bit set and is a build option, added landing direction offset parameter for when you want the landing to be different direction than takeoff

PS: once merged I have a PR to add a mode param to allow the direction to be set on arm if a compass is in use instead of autotakeoffs, or to allow setting takeoff direction in FBWA/MANUAL/STABILIZE takeoffs in addition to autotakeoffs

timtuxworth reviewed (commented):
Love this idea - some suggestions.

Hwurzburg commented:
@robustini many people never use MP after initial setup and arent comfortable with mission planning....I never use missions other than autotakeoff/do-land-start/landing sequence with my mission selector app to select land direction for my field based on wind direction....
this is the C++ implementation of my UniversalAutoLand lua script which I use when traveling and just stop at a location and fly...no laptop used..

Hwurzburg commented on ArduPlane/mode_autoland.cpp:
```diff
@@ -0,0 +1,104 @@
+const AP_Param::GroupInfo ModeAutoLand::var_info[] = {
+    // @Param: WP_ALT
+    // @DisplayName: Final approach WP altitude
+    // @Description: This is the target altitude for final approach waypoint
```
virtually every other alt param we have implies above home,but it doesnt hurt, so done

timtuxworth commented:
Should this be disabling fences (autoenabled, and floor) when doing a landing?

Hwurzburg commented:
It does a NAV_LAND....so whatever it does or does not

Hwurzburg commented on ArduPlane/mode_autoland.cpp:
```diff
@@ -0,0 +1,104 @@
+    plane.next_WP_loc = home;
+    uint16_t bearing_cd = wrap_360((plane.takeoff_state.takeoff_initial_direction + 180));
+    plane.next_WP_loc.offset_bearing(bearing_cd, final_wp_dist);
+    plane.next_WP_loc.alt = home.alt + final_wp_alt*100;
```
done

peterbarker commented on Tools/autotest/arduplane.py:
```diff
@@ -4200,6 +4200,19 @@ def FlyEachFrame(self):
+    def AutoLandMode(self):
+        '''Test AUTOLAND mode'''
+        self.customise_SITL_commandline(["--home", "-35.362938,149.165085,585,173"])
+        self.change_mode("TAKEOFF")
+        self.wait_ready_to_arm()
+        self.arm_vehicle()
```
```suggestion
        self.context_collect('STATUSTEXT')
        self.takeoff(mode='TAKEOFF')
```

peterbarker commented on Tools/autotest/arduplane.py:
```diff
@@ -4200,6 +4200,19 @@ def FlyEachFrame(self):
+        self.change_mode("TAKEOFF")
+        self.wait_ready_to_arm()
+        self.arm_vehicle()
+        self.wait_text("Takeoff initial direction")
```
```suggestion
        self.wait_text("Takeoff initial direction", check_context=True)
```

Hwurzburg commented on ArduPlane/events.cpp:
```diff
@@ -145,6 +146,8 @@ void Plane::failsafe_long_on_event(enum failsafe_state fstype, ModeReason reason
         } else if (g.fs_action_long == FS_ACTION_LONG_AUTO) {
             set_mode(mode_auto, reason);
+        } else if (g.fs_action_long == FS_ACTION_LONG_AUTOLAND) {
+            set_mode(mode_autoland, reason);
```
it does in the mode

IamPete1 replied:
Set mode can stay the same, the failsafe case is the only one that should fall-through to RTL. RC or GCS mode changes should just fail. This is how copter does it:

https://github.com/ArduPilot/ardupilot/blob/da4fee5a9a94511ce778a3b740990f8bcf0e12fb/ArduCopter/events.cpp#L424-L437

It trys auto RTL, if that fails it trys RTL, if that fails it trys land.

Hwurzburg commented on ArduPlane/mode_autoland.cpp:
```diff
@@ -0,0 +1,119 @@
+bool ModeAutoLand::_enter()
+{     
+    //if do_land_start exists, jump to it
+    if( (plane.mission.contains_item(MAV_CMD_DO_LAND_START)) && (plane.arming.is_armed())) {
+        if (plane.have_position && plane.mission.jump_to_landing_sequence(plane.current_loc)) {
+            // switch to AUTO
+            plane.mission.set_force_resume(true);
+            if (plane.set_mode(plane.mode_auto, ModeReason::FIXED_WING_AUTOLAND)) {
+                 return true;
+            }
+        }
+    }
```
dont know what you mean, Tridge wanted this

Hwurzburg commented on ArduPlane/mode_autoland.cpp:
```diff
@@ -0,0 +1,119 @@
+    if (!plane.takeoff_state.takeoff_direction_initialized) {
+        gcs().send_text(MAV_SEVERITY_WARNING, "Takeoff initial direction not set,must use autotakeoff");
+        gcs().send_text(MAV_SEVERITY_WARNING, "Changing to RTL"); 
+        plane.set_mode(plane.mode_rtl, ModeReason::FIXED_WING_AUTOLAND);
```
done

Hwurzburg commented on ArduPlane/mode_autoland.cpp:
```diff
@@ -0,0 +1,119 @@
+    cmd.content.location = plane.next_WP_loc;
+    plane.start_command(cmd);
+    land_started = false;
+    gcs().send_text(MAV_SEVERITY_WARNING, "AutoLanding");
```
okay

Hwurzburg commented on ArduPlane/mode_autoland.cpp:
```diff
@@ -0,0 +1,119 @@
+#if HAL_QUADPLANE_ENABLED
+    if (quadplane.available()) {
+        gcs().send_text(MAV_SEVERITY_WARNING, "AutoLand is fixed wing only,changing to RTL"); 
+        plane.set_mode(plane.mode_rtl, ModeReason::FIXED_WING_AUTOLAND);
```
there are 3 reasons to fail...two result in switch to RTL...its a lot less code to do the switch in the enter rather than hack around system.cpp and change the architecture of set_mode, IMHO

IamPete1 replied:
Well, here your assuming that the user wants RTL. They might prefer to do something else. If you refuse the mode enter they can decide. Maybe they want QRTL or QLAND instead of RTL. These eventualities result in us having to add more Q_OPTIONS in the future, better to just reject the mode change.

Hwurzburg replied:
okay no mode changes in enter

Hwurzburg commented on ArduPlane/mode.h:
```diff
@@ -835,6 +836,45 @@ class ModeTakeoff: public Mode
+private:
+
+
+
```
done

Hwurzburg commented on ArduPlane/Parameters.h:
```diff
@@ -100,7 +100,7 @@ class Parameters {
         k_param_log_bitmask,
         k_param_BoardConfig,
-        k_param_rssi_range,     // unused, replaced by rssi_ library parameters
+        k_param_mode_autoland,     // was rssi_range
```
moved it

tridge commented on ArduPlane/mode_autoland.cpp:
```diff
@@ -0,0 +1,112 @@
+bool ModeAutoLand::_enter()
+{ 
+    //if do_land_start exists, jump to it
+    if( (plane.mission.contains_item(MAV_CMD_DO_LAND_START)) && (plane.arming.is_armed())) {
```
i'd prefer just to remove this

Hwurzburg replied:
removed

Hwurzburg commented on ArduPlane/mode.h:
```diff
@@ -835,6 +836,45 @@ class ModeTakeoff: public Mode
+
+protected:
+    bool _enter() override;
+    AP_Mission::Mission_Command cmd {};
```
done

Hwurzburg commented:
Added AUTOLAND_DIR_OFS param to rotate the landing direction from takeoff direction as requested....

Test flown today without any issues...

peterbarker commented on ArduPlane/mode_autoland.cpp:
```diff
@@ -0,0 +1,114 @@
+    
+    //takeoff direction must be set and must not be a quadplane, otherwise since flying switch to RTL so this can be used as FS action
+#if HAL_QUADPLANE_ENABLED
+    if (quadplane.available()) {
+        gcs().send_text(MAV_SEVERITY_WARNING, "AutoLand is fixed wing only mode"); 
```
We should permit the use of this mode if the Q_OPTION bit is set to allow fixed-wing landings.

Hwurzburg replied:
fine by me and easy to do, but @tridge was the one who asked me to prevent Qplanes from using this

peterbarker replied:
He reminded me the Q_OPTION bit existed :-)

Hwurzburg replied:
@peterbarker thanks for the flight testing! added the ability to use this mode in QPlane if option is set....also landing auto-flaps WAS working, although incorrectly: they were activated immediately upon mode entry while flying to final waypoint...the last commit (unsquashed so changes are easily reviewed) fixes this....SITL tested for flaps
my namespace mojo is weak and still get a compile error on the Qoption addition even adding the QPlane friend class...need help fixing that

tridge commented:
@Hwurzburg Peter and I tested yesterday, looked pretty good!
One question is if flaps were correctly deployed in landing, @peterbarker can you check the bin log?

Hwurzburg commented:
@tridge read my comment above yours

Hwurzburg commented:
added all requests and fixed auto-flaps to be active only in final landing segment instead of entire mode duration...should be good to go now

IamPete1 commented on ArduPlane/Plane.h:
```diff
@@ -454,6 +459,11 @@ class Plane : public AP_Vehicle {
         float throttle_lim_min;
         uint32_t throttle_max_timer_ms;
         uint32_t level_off_start_time_ms;
+        // Good candidate for keeping the initial time for TKOFF_THR_MAX_T.
+#if MODE_AUTOLAND_ENABLED
+        int32_t takeoff_initial_direction; //deg
+        bool takeoff_direction_initialized;
+#endif
```
Comment is out of date and int32 is a bit unexpected for direction, why not float?
```suggestion
#if MODE_AUTOLAND_ENABLED
        struct {
            float heading; // deg
            bool initialized;
        } initial_direction;
#endif
```
You could also add a little structure to keep the variable names shorter.

Hwurzburg replied:
done

IamPete1 commented on ArduPlane/events.cpp:
```diff
@@ -145,6 +149,12 @@ void Plane::failsafe_long_on_event(enum failsafe_state fstype, ModeReason reason
             set_mode(mode_auto, reason);
+#if MODE_AUTOLAND_ENABLED
+        } else if (g.fs_action_long == FS_ACTION_LONG_AUTOLAND) {
+            if(!set_mode(mode_autoland, reason)) {
```
```suggestion
            if (!set_mode(mode_autoland, reason)) {
```

Hwurzburg replied:
done

IamPete1 commented on ArduPlane/events.cpp:
```diff
@@ -194,14 +204,23 @@ void Plane::failsafe_long_on_event(enum failsafe_state fstype, ModeReason reason
             set_mode(mode_auto, reason);
+#if MODE_AUTOLAND_ENABLED
+        } else if (g.fs_action_long == FS_ACTION_LONG_AUTOLAND) {
+            if(!set_mode(mode_autoland, reason)) {
```
```suggestion
            if (!set_mode(mode_autoland, reason)) {
```

Hwurzburg replied:
done

IamPete1 commented on ArduPlane/events.cpp:
```diff
@@ -194,14 +204,23 @@ void Plane::failsafe_long_on_event(enum failsafe_state fstype, ModeReason reason
+            }            
         } else if (g.fs_action_long == FS_ACTION_LONG_RTL) {
             set_mode(mode_rtl, reason);
+#endif
```
Should this be before the else if above?

Hwurzburg replied:
yes...my bad

IamPete1 commented on ArduPlane/mode_autoland.cpp:
```diff
@@ -0,0 +1,118 @@
+    }
+#endif
+    if (!plane.takeoff_state.takeoff_direction_initialized) {
+        gcs().send_text(MAV_SEVERITY_WARNING, "Takeoff initial direction not set,must use autotakeoff");
```
```suggestion
        gcs().send_text(MAV_SEVERITY_WARNING, "Takeoff initial direction not set, must use autotakeoff");
```

Hwurzburg replied:
done

IamPete1 commented on ArduPlane/mode_autoland.cpp:
```diff
@@ -0,0 +1,118 @@
+    const Location &home = ahrs.get_home();
+    plane.set_target_altitude_current();
+    plane.next_WP_loc = home;
+    uint16_t bearing_cd = wrap_360_cd((plane.takeoff_state.takeoff_initial_direction + 180)*100);
+    plane.next_WP_loc.offset_bearing(bearing_cd * 0.01f, final_wp_dist);
```
```suggestion
    plane.next_WP_loc.offset_bearing(plane.takeoff_state.takeoff_initial_direction, -final_wp_dist);
```
Negative offset distance remvoe the step adding 180.

Hwurzburg replied:
no this is required to put the waypoint behind the home waypoint

IamPete1 replied:
Using a negative offset is the same as moving the bearing by 180 degrees.

Hwurzburg replied:
actually that does not work...the offset method apparently expects -180 t0 +360 degs for the arguement ...attempted in SITL and it wraps at -180 so it wouldn't really be saving much since a  conversion would have to be done anyway I think

IamPete1 replied:
You should not have to wrap, the function will take any value.

Hwurzburg replied:
not -270....it wraps to -90 which offsets the wrong direction...but will fix the cd vs deg as suggested

IamPete1 reviewed (changes requested):
Basically looks good, just a couple of things.

timtuxworth commented on ArduPlane/mode_autoland.cpp:
```diff
@@ -0,0 +1,118 @@
+    }
+#endif
+    if (!plane.takeoff_state.takeoff_direction_initialized) {
+        gcs().send_text(MAV_SEVERITY_WARNING, "Takeoff initial direction not set,must use autotakeoff");
```
This message is a bit long. I'm sure it's going to wrap on Mission Planner. Make two messages?

Hwurzburg replied:
it will wrap in the MP messages tab, like many other messages...but making it two will make the first one not show in HUB and OSD where last GCS send text always wins

IamPete1 replied:
Yes, lets not try and fix Mission Planner issues in AP. It should re-assemble messages if it needs to.

timtuxworth commented on ArduPlane/mode_autoland.cpp:
```diff
@@ -0,0 +1,118 @@
+    //takeoff direction must be set and must not be a quadplane, otherwise since flying switch to RTL so this can be used as FS action
+#if HAL_QUADPLANE_ENABLED
+    if (quadplane.available() && !quadplane.option_is_set(QuadPlane::OPTION::ALLOW_FW_LAND)) {
+        gcs().send_text(MAV_SEVERITY_WARNING, "Option not set to allow fixed wing autoland"); 
```
How about "fixed wing autoland not enabled" (shorter)

Hwurzburg replied:
change to "Fixed wing autoland option disabled"

IamPete1 commented on ArduPlane/mode_autoland.cpp:
```diff
@@ -0,0 +1,118 @@
+    const Location &home = ahrs.get_home();
+    plane.set_target_altitude_current();
+    plane.next_WP_loc = home;
+    uint16_t bearing_cd = wrap_360_cd((plane.takeoff_state.initial_direction.heading + 180)*100);
+    plane.next_WP_loc.offset_bearing(bearing_cd * 0.01f, final_wp_dist);
```
At least don't go to cd and back.
```suggestion
    const float bearing = wrap_360(plane.takeoff_state.initial_direction.heading + 180);
    plane.next_WP_loc.offset_bearing(bearing, final_wp_dist);
```

Hwurzburg replied:
fair enough...left over from previous cd based stuff

tridge commented:
@peterbarker need the bin log of your tests

peterbarker commented:
Log from testing on crnch: https://drive.google.com/file/d/1Hwq9iGn1AlVooi7GgTMGQud1k-Sf-NF-/view?usp=sharing

Hwurzburg commented:
this looks 100% correct and matches what master would do if that final waypoint and nav_land waypoint were the content of a do_land_start sequence....verified that in master in SITL

IamPete1 reviewed (changes requested):
New mode needs adding to this array for the available modes message.

https://github.com/ArduPilot/ardupilot/blob/5a8bc161709aba97a8b7a6da3ce0ec1a49df2718/ArduPlane/GCS_Mavlink.cpp#L1592

Sorry, this was added after you started your PR. I can push a commit for it if you prefer.

Hwurzburg commented:
@IamPete1 if you could, it would help...I tried adding it and get an error message that MODE_AUTOLAND_ENABLED is not defined around the addition

IamPete1 commented:
> @IamPete1 if you could, it would help...I tried adding it and get an error message that MODE_AUTOLAND_ENABLED is not defined around the addition

@Hwurzburg  Fix here: https://github.com/IamPete1/ardupilot/tree/autoland I also rebased.

Hwurzburg commented:
@IamPete1 fix added...thanks....can you approve now?

IamPete1 reviewed (approved):
Looks good.

tridge commented:
I think this needs a base leg of 1/4 of the approach distance, like this:
![image](https://github.com/user-attachments/assets/b48346c2-42fe-4ee0-8891-b282a41a3be9)
the base leg will keep the aircraft in the part of the field the pilot is flying in, otherwise we end up with a path like the black line above which is what we got with the bixler

tridge commented:
@Hwurzburg I've added a base leg to the PR. This is with the base leg:
![image](https://github.com/user-attachments/assets/798d88b0-23da-4839-b577-65dd846211b0)

IamPete1 commented:
> I think this needs a base leg of 1/4 of the approach distance, like this: ![image](https://private-user-images.githubusercontent.com/831867/398811938-b48346c2-42fe-4ee0-8891-b282a41a3be9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzUzMDI4MjcsIm5iZiI6MTczNTMwMjUyNywicGF0aCI6Ii84MzE4NjcvMzk4ODExOTM4LWI0ODM0NmMyLTQyZmUtNGVlMC04ODkxLWIyODJhNDFhM2JlOS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjI3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIyN1QxMjI4NDdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kMzEwYjgxMDYxYjA2ODUwYTU3YjQ1ODMwNGYyNDk4MGI2Y2I3M2QwYWJjOWE4OTJjYjNiZDc5NmVlYTA5ZDNkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.en_a6XZY2jcjFTiyeyul9o3INnoH2tJkJzQaR9w2GyY) the base leg will keep the aircraft in the part of the field the pilot is flying in, otherwise we end up with a path like the black line above which is what we got with the bixler

I don't understand how the shown path related to the red(?) line which I guess was the points this added as it was when it was flown? If it was no where near the set path then how does adding the extra leg help?

Hwurzburg commented:
@IamPete1 the base leg addition was to minimize the overshoot of the final approach waypoint by forcing the approach to it to be no more than 90 degs before the turn onto final....otherwise the approach to the final approach waypoint could have been from 180degs to landing direction resulting in larger overshoot and more track error to correct as it lined back up on final leg...
depending on plane config, roll max, and wp-rad you can get a large overshoot otherwise without base leg and Tridge has fences to the south of the landing approach that could breach....adding a base leg minimizes the overshoot amounts which can be further tunedd with wp_rad,max bank, etc.

@trdige SITL looks good....will test fly this weekend

tridge commented:
@Hwurzburg tested on Peters plane, worked well:
![Uploading image.png…]()

tridge commented:
> I don't understand how the shown path related to the red(?) line which I guess was the points this added as it was when it was flown? If it was no where near the set path then how does adding the extra leg help?

there were a few reasons it didn't do what it should do without the base leg:
 - the up to 180 degree turnaround as Henry mentioned
 - it wasn't flying all the way to the autoland start point as it was using WP_RADIUS, and as there is no mission it can't know the turn angle for the next wp, so can't judge when to turn
 - it uses a lot (sometimes nearly all) of the autoland distance getting back onto the landing track
 - the geometry meant it always did the landing from the opposite side of the field that the pilot was flying in (as it guaranteed an overshoot due to the above reasons)

A normal fixed wing landing has a base leg for all size aircraft. The only time you can really avoid it is if you happen to start the landing sequence when already lined up for the landing. We could add a special case for that if we think it is important

Hwurzburg commented:
During testing I found a corner case bug.....multiple mode takeoffs in a row sometimes triggers the clearing of takeoff_state structure...before I added the bearing into that structure I had accounted for this...but when we went to a substructure including the bearing, it was not included in code that prevents its clearing....fixed....I did not squash, but will clean the commit tree once we agree to add @tridge  base leg....(which I do)

LupusTheCanine commented:
I personally use LOITER_TO_ALT with tangent exit with great success for landing entry with great success.

I did a bit more SITL testing and it has ugly entry for approaches that are 90° to the wrong side. but it can be fixed by preceding it with LOITER_TURNS set to 0.125 (45°).

![image](https://github.com/user-attachments/assets/07b2ea06-3a4d-4b2c-81d8-a40acf25dba4)

Here is example waypoint list for EPWA.
```
3	DO_LAND_START	0	0	0	0	0	0	50	Relative	X			0	0	0	0
4	LOITER_TURNS	0.125	0	50	1	52.1719446	20.9432673	35	Relative	X			0.0	0.0	273.3	154
5	LOITER_TO_ALT	0	50	0	1	52.1719446	20.9432673	35	Relative	X			NaN	NaN	0.0	180
6	LAND	0	0	0	1	52.1697995	20.9527194	0	Relative	X			-5.1	-2.9	688.2	110
```
